test/jit/test_await.py

```diff
@@ -0,0 +1,331 @@
+
+        @torch.jit.script
+        def main(x: Tensor) -> Tensor:
+            awaw = fn(x)
+            return torch.jit._awaitable_wait(torch.jit._awaitable_wait(awaw))
+
+        inp = torch.zeros(2)
+
+        sm = torch.jit.script(main)
+        out = main(inp)
+        script_out = sm(inp)
+        self.assertTrue(torch.allclose(script_out, out))
+
+    def test_eager_await_non_scriptable(self):
+        # Tree type can not be compiled (Recursive type)
+        class Tree(object):
+            def __init__(self, v):
+                self.parent = torch.jit.annotate(Optional[Tree], None)
+                self.v = v
+        make_global(Tree)
+
+        def delayed(t: Tree):
+            t.v = t.v + 1
+            return t
+
+        aw = torch.jit._awaitable(delayed, Tree(2))
+        t = torch.jit._awaitable_wait(aw)
+        self.assertTrue(t.v == 3)
+
+    def test_await_isinstance(self):
+        def delayed(x: Tensor) -> Tensor:
+            return 2 * (x + 1)
+
+        @torch.jit.script
+        def main(x: Tensor) -> Tensor:
+            aw = torch.jit._awaitable(delayed, x)
+            assert isinstance(aw, torch.jit.Await)
+            return torch.jit._awaitable_wait(aw)
+
+        inp = torch.zeros(2)
+
+        sm = torch.jit.script(main)
+        out = main(inp)
+        script_out = sm(inp)
+        self.assertTrue(torch.allclose(script_out, out))
+
+    def test_await_eager_lazy(self):
+        def delayed(x: Tensor) -> Tensor:
+            return 2 * (x + 1)
+        t = torch.ones(2, dtype=torch.int64)
+        aw = torch.jit._awaitable(delayed, t)
+        self.assertTrue(isinstance(aw, torch._C.Await))
+        self.assertTrue(t.dtype == aw.dtype)
+
+    def test_await_out_of_interpreter(self):
+        def delayed(x: Tensor) -> Tensor:
+            return 2 * (x + 1)
+
+        @torch.jit.script
+        def main(x: Tensor) -> Await[Tensor]:
+            aw = torch.jit._awaitable(delayed, x)
+            assert isinstance(aw, torch.jit.Await)
+            return aw
+
+        inp = torch.zeros(2)
+
+        sm = torch.jit.script(main)
+        out_aw = main(inp)
+        out = torch.jit._awaitable_wait(out_aw)
+
+        script_out_aw = sm(inp)
+        script_out = torch.jit._awaitable_wait(script_out_aw)
+
+        self.assertTrue(torch.allclose(script_out, out))
+
+    def test_jit_trace(self):
+        def gap(x: Tensor):
+            return torch.relu(x) + torch.sin(x)
+
+        def delayed(x: Tensor) -> Tensor:
+            return 2 * (torch.cos(x) + 1)
+
+        def main(x: Tensor, y: Tensor) -> Tensor:
+            aw = torch.jit._awaitable(delayed, x)
+
+            assert isinstance(aw, torch.Await)
+            z = gap(y)
+            k = torch.jit._awaitable_wait(aw)
+            return y + k
+
+        inp = torch.randn(2)
+        tm = torch.jit.trace(main, (inp, inp))
+        inp_check = torch.ones(2)
+        self.assertEqual(main(inp_check, inp_check), tm(inp_check, inp_check))
```

Comment:
a few other things to test (ignore if they are already tested, I may have missed them)
* awaitable that returns multiple outputs
* parse_ir ([examples](https://github.com/search?q=repo%3Apytorch%2Fpytorch+parse_ir+path%3A%2F%5Etest%5C%2Fjit%5C%2F%2F&type=code&l=test%2Fjit%2F)) - tbh it's not a dealbreaker if this doesn't work, but would be good to have signal on it if it does.
* serialization ([examples](https://github.com/pytorch/pytorch/blob/fbafcecf8de933176f55e5dd2d56169d16bed502/test/jit/test_save_load.py#L209)]

Reply:
oh also - from what I remember is it true that you want to be able to pass Awaits as inputs to functions? If so can you add a test for that as well?